src/point_one/rtcm/rtcm_framer.cc

```diff
@@ -0,0 +1,481 @@
+  bool crc_check_needed = false;
+  if (state_ == State::SYNC) {
+    VLOG(4) << "Searching for sync byte. [byte=" << HexPrintableInteger(byte)
+            << "]";
+    if (byte == RTCM3_PREAMBLE) {
+      VLOG(4) << "Found sync byte 0.";
+      state_ = State::HEADER;
+    } else {
+      --next_byte_index_;
+    }
+  }
+  // Search for a message header.
+  else if (state_ == State::HEADER) {
+    VLOG(4) << "Received " << next_byte_index_ << "/" << RTCM_HEADER_BYTES
+            << " header bytes. [byte=" << HexPrintableInteger(byte) << "]";
+
+    // Check if the header is complete.
+    if (next_byte_index_ == RTCM_HEADER_BYTES) {
+      // Compute the full message size. If the message is too large to fit in
+      // the buffer, we cannot parse it. Otherwise, start collecting the
+      // message payload.
+      //
+      // Note that while we compute the current_message_size_ here, we
+      // intentionally do the "too big" check below with the payload size. That
+      // way we implicitly handle cases where the payload is large enough to
+      // cause current_message_size_ to overflow. Normally, this won't happen
+      // for legit packets that are just too big for the user's buffer, but it
+      // could happen on a bogus header if we find the preamble randomly in an
+      // incoming byte stream. The buffer capacity is always
+      // >=RTCM_OVERHEAD_BYTES, so the subtraction will never be negative.
+      /**
+       * Transport header (big endian):
+       *   Preamble = 8 bits
+       *   Reserved = 6 bits
+       *   Message Length = 10 bits
+       */
+      uint16_t header_byte_1_2_le = EndianSwap16(buffer_ + 1);
+      uint16_t payload_size_bytes = header_byte_1_2_le & 0x3F;
+      current_message_size_ = payload_size_bytes + RTCM_OVERHEAD_BYTES;
+      VLOG(3) << "Header complete. Waiting for payload. [payload_size="
+              << payload_size_bytes << " B]";
+      if (current_message_size_ <= capacity_bytes_ &&
+          current_message_size_ <= RTCM_MAX_SIZE_BYTES) {
+        state_ = State::DATA;
+      } else {
+        error_count_++;
+        if (quiet) {
+          VLOG(2) << "Message too large for buffer. [size="
+                  << current_message_size_
+                  << " B (payload=" << payload_size_bytes
+                  << " B), buffer_capacity=" << capacity_bytes_
+                  << " B (max_payload=" << capacity_bytes_ - RTCM_OVERHEAD_BYTES
+                  << " B)]";
+        } else {
+          LOG(WARNING) << "Message too large for buffer. [size="
+                       << current_message_size_
+                       << " B (payload=" << payload_size_bytes
+                       << " B), buffer_capacity=" << capacity_bytes_
+                       << " B (max_payload="
+                       << capacity_bytes_ - RTCM_OVERHEAD_BYTES << " B)]";
+        }
+
+        state_ = State::SYNC;
+        return -1;
+      }
+    }
+  }
+  // Collect the message payload.
```

Comment:
Might be worth saying "payload and CRC". Wasn't obvious that `current_message_size_` includes the CRC.